File: code/SWE2-Project/Animal Tracker/src/main/java/edu/westga/cs3211/animaltracker/model/Project.java
package edu.westga.cs3211.animaltracker.model;

import java.util.List;

/**
 * The Project Class.
 *
 * @author mrocker1
 */
public class Project {

    private final String name;
    private final List<Animal> animals;
    private final List<User> users;
    private int id;

    /**
     * Instantiates a new Project.
     *
     * @param users   the users
     * @param name    the name of the project
     * @param animals the animals in the project
     * @pre users != null && name != null or empty && animals != null
     * @post getUsers().equals(users) &&
     * getName.equals(name) &&
     * getAnimals.equals(animals) &&
     * getId = random project id.
     */
    public Project(List<User> users, String name, List<Animal> animals) {
        if (users == null) {
            throw new IllegalArgumentException("Users cannot be null");
        }

        if (name == null) {
            throw new IllegalArgumentException("Name cannot be null");
        }

        if (name.isBlank()) {
            throw new IllegalArgumentException("Name cannot be empty");
        }

        if (animals == null) {
            throw new IllegalArgumentException("Animals cannot be null");
        }
        this.name = name;
        this.animals = animals;
        this.users = users;
        this.id = DataStorage.getNextProjectId();
        DataStorage.getProjects().put(this.getId(), this);
    }

    /**
     * Instantiates a new Project.
     *
     * @param users   the users
     * @param name    the name of the project
     * @param animals the animals in the project
     * @param id the id of the user
     * @pre users != null && name != null or empty && animals != null
     * @post getUsers().equals(users) &&
     * getName.equals(name) &&
     * getAnimals.equals(animals) &&
     * getId = random project id.
     */
    public Project(List<User> users, String name, List<Animal> animals, int id) {
        this(users, name, animals);
        this.id = id;
        System.out.println("success");
    }

    /**
     * Gets the Name of the Project.
     *
     * @return the name of the Project
     */
    public String getName() {
        return this.name;
    }

    /**
     * Gets the List of Animals.
     *
     * @return the List of Animals
     */
    public List<Animal> getAnimals() {
        return this.animals;
    }

    /**
     * Gets the List of Users.
     *
     * @return the List of Users
     */
    public List<User> getUsers() {
        return this.users;
    }

    /**
     * Adds an animal to the List of Animals.
     *
     * @param animal the Animal
     * @pre animal != null && !animals.contains(animal)
     * @post animals.size() = animals.size()@prev + 1 &&
     * containsAnimal(animal) == true
     */
    public void addAnimal(Animal animal) {
        if (animal == null) {
            throw new IllegalArgumentException("animal cannot be null");
        }
        if (this.containsAnimal(animal)) {
            throw new IllegalArgumentException("animal already exists");
        }
        this.animals.add(animal);
    }

    /**
     * Checks if the Animal is in the List.
     *
     * @param animal the Animal
     * @return true if animal is in the List, false otherwise.
     * @pre animal != null
     * @post none
     */
    public boolean containsAnimal(Animal animal) {
        return this.animals.contains(animal);
    }

    /**
     * Checks if an AnimalClass is in the List.
     *
     * @param animalClass the animalClass
     * @return true if there is an animal of the AnimalClass, false otherwise
     * @pre animalClass != null
     * @post none
     */
    public boolean containsAnimalType(AnimalClass animalClass) {
        if (animalClass == null) {
            throw new IllegalArgumentException("animalClass cannot be null");
        }
        return this.animals.stream().anyMatch(animal -> animal.getAnimalClass().equals(animalClass));
    }

    /**
     * Returns a List of Animals of a specific Animal Class.
     *
     * @param animalClass The Animal Class
     * @return The List of Animals of an Animal Class.
     * @pre animalClass != null
     * @post none
     */
    public List<Animal> getAnimalsByType(AnimalClass animalClass) {
        if (animalClass == null) {
            throw new IllegalArgumentException("animalClass cannot be null");
        }
        return this.animals.stream().filter(animal -> animal.getAnimalClass().equals(animalClass)).toList();
    }

    /**
     * Adds a User to the Project.
     *
     * @param user the User
     * @pre user != null
     * @post users.size() = users.size()@prev + 1
     */
    public void addUser(User user) {
        if (user == null) {
            throw new IllegalArgumentException("user cannot be null");
        }
        if (this.containsUser(user)) {
            throw new IllegalArgumentException("user already exists");
        }
        this.users.add(user);
    }

    /**
     * Checks if the User is in the List.
     *
     * @param user the User
     * @return true if user is in the List, false otherwise.
     * @pre none
     * @post none
     */
    public boolean containsUser(User user) {
        return this.users.contains(user);
    }

    /**
     * Removes a User from the Project.
     *
     * @param user the User
     * @pre user != null
     * @post users.size() = users.size()@prev - 1 if user was in the list
     */
    public void removeUser(User user) {
        if (user == null) {
            throw new IllegalArgumentException("user cannot be null");
        }
        if (!this.containsUser(user)) {
            throw new IllegalArgumentException("user does not exist in project");
        }
        this.users.remove(user);
    }

    /**
     * Gets the project id.
     *
     * @return the project id
     */
    public int getId() {
        return this.id;
    }
}

File: code/SWE2-Project/Animal Tracker Server/src/model/Project.py

class Project:
    def __init__(self, name, users, animals, project_id):
        """
        Initializes a new project.
        :param name: The name of the project.
        :param users: The users of the project.
        :param animals: The animals of the project.
        :param project_id: The ID of the project.
        """
        if name is None:
            raise Exception("Project name cannot be None")
        if users is None:
            raise Exception("Project users cannot be None")
        if project_id < 0:
            raise Exception("Project project_id cannot be negative")
        if name.strip() == "":
            raise Exception("Project name cannot be an empty string")
        if animals is None:
            raise Exception("Project animals cannot be None")
        self._name = name
        self._users = users
        self._animals = animals
        self._id = project_id

    def get_name(self):
        """
        Gets the name of the project.
        :return: The name of the project.
        """
        return self._name
    def get_users(self):
        """
        Gets the users in the project.
        :return: The users in the project.
        """
        return self._users
    def get_animals(self):
        """
        Gets the animals in the project.
        :return: The animals in the project.
        """
        return self._animals

    def get_id(self):
        """
        Gets the id of the project.
        :return: The id of the project.
        """
        return self._id
    def set_name(self, name):
        """
        Sets the name of the project with a new name.
        :param name: The new name of the project.
        """
        if name is None:
            raise Exception("Project name cannot be None")
        if name.strip() == "":
            raise Exception("Project name cannot be an empty string")
        self._name = name

    def set_id(self, id):
        """
        Sets the id of the project with a new id.
        :param id: The new id of the project.
        """
        if id is None:
            raise Exception("Project id cannot be None")
        if id < 0:
            raise Exception("Project id cannot be negative")
        self._id = id

    def add_user(self, user):
        """
        Adds a user to the project.
        :param user: The user to add.
        """
        if user is None:
            raise Exception("User cannot be None")
        if self.contains_user(user):
            raise Exception("User already exists")
        self._users.add(user)

    def contains_user(self, user):
        """
        Checks if the user is in the project.
        :param user: The user to check.
        """
        return user in self._users

    def add_animal(self, animal):
        """
        Adds a new animal to the project.
        :param animal: The animal to add.
        """
        if animal is None:
            raise Exception("Animal cannot be None")
        if self.contains_animal(animal):
            raise Exception("Animal already exists")
        self._animals.add(animal)
    def contains_animal(self, animal):
        """
        Checks if the project contains the animal.
        :param animal: The animal to check.
        :return: True if the project contains the animal.
        """
        return animal in self._animals

    def remove_user(self, user):
        """
        Removes a user from the project.
        :param user: The user to remove.
        """
        if user is None:
            return
        self._users.remove(user)